[Ansor][AutoTVM v2.0] Phase 1: Add annotation/compute_at/compute_root/compute_inline steps

jcf94 opened the pull request:
For full upstream plan, see [Ansor RFC](https://discuss.tvm.ai/t/rfc-ansor-an-auto-scheduler-for-tvm-autotvm-v2-0/7005/32).

In this PR, we bring annotation/compute_at/compute_root/compute_inline steps for Ansor auto_scheduler.

To record the stage/iterator attach relations, we introduced a new structure AttachMap in State.

cc @tqchen @merrymercy @FrozenGene @minminsun @comaniac

junrushao1994 commented:
Just noticed that #6077 is also titled as "part 1", shall we make one of them part 2 instead?

comaniac commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -143,12 +265,67 @@ void State::DoReorderStep(const ReorderStep& step) {
+void State::DoComputeRootStep(const ComputeRootStep& step) {
+  const Stage& stage = operator->()->stages[step->stage_id];
+
+  // After compute_at, we don't know the accurate length information any more
+  // If we do want to know the accurate lengths, we can call ComputeDAG::InferBound
```
ditto.

comaniac commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -143,12 +265,67 @@ void State::DoReorderStep(const ReorderStep& step) {
+void State::DoComputeInlineStep(const ComputeInlineStep& step) {
+  const Stage& stage = operator->()->stages[step->stage_id];
+
+  // CHECK the validity of compute_inline
```
s/CHECK/Check

comaniac commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -143,12 +265,67 @@ void State::DoReorderStep(const ReorderStep& step) {
+    CHECK_EQ(operator->()->attach_map->iter_to_attached_stages.count(
+                 std::make_pair(step->stage_id, i)),
+             0)
+        << "Invalid compute_inline: Because there are some other stages "
```
Remove "Because"

comaniac commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -210,6 +387,16 @@ Array<Iterator> State::DoSplitStepCommon(int stage_id, int iter_id,
                      Stage(stage->op, stage->op_type, new_iters, stage->compute_at, stage->attrs));
   pstate->concrete &= concrete;
 
+  // We have to update the iterator relations in attach map, these two vectors keep the replacement
+  // mapping
```
```suggestion
  // Use two vectors to represent the iterator relation before and after the split
  // in order to update the attach_map
```

We may also need to mention in the split function defintion in `loop_state.h` that if the attached target iterator is split, then the innermost split iterator would be the new attached target.

comaniac commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -264,6 +462,38 @@ Iterator State::DoFuseStep(const FuseStep& step) {
   pstate->stages.Set(stage_id,
                      Stage(stage->op, stage->op_type, new_iters, stage->compute_at, stage->attrs));
 
+  // We have to update the iterator relations in attach map, these two vectors keep the replacement
+  // mapping
```
ditto.

comaniac commented on tests/python/unittest/test_auto_scheduler_loop_state.py:
```diff
@@ -61,5 +61,79 @@ def test_split_fuse_reorder():
+    s0.compute_inline(bn_mul)
+    s0.compute_inline(bias_add)
+    s0.compute_at(conv, relu, s0[relu].iters[2])
+    print(s0)
```
Remove this line.

comaniac commented on tests/python/unittest/test_auto_scheduler_loop_state.py:
```diff
@@ -61,5 +61,79 @@ def test_split_fuse_reorder():
+    s0.compute_inline(bias_add)
+    s0.compute_at(conv, relu, s0[relu].iters[2])
+    print(s0)
+    assert str(s0) == \
```
I feel that it's not a good idea to check the correctness using printed strings...

jcf94 replied:
I agree, but failed to figure out a better way here. At least the output string of state will not be Inferenced by other TVM parts out than `auto_scheduler` dir.
@merrymercy Any comment?

merrymercy replied:
I have no idea either

jcf94 replied:
Update the UT to check each iterator

comaniac commented on tests/python/unittest/test_auto_scheduler_loop_state.py:
```diff
@@ -61,5 +61,79 @@ def test_split_fuse_reorder():
     assert s1[C].iters[4].range.extent == 8
     assert s1[C].iters[5].range.extent == 2
 
+    s1.parallel(C, j1)
+    s1.unroll(C, j2)
+    s1.vectorize(C, j3)
+    s1.bind(C, i1, "blockIdx.x")
+    s1.bind(C, i2, "vthread")
+    s1.bind(C, i3, "threadIdx.y")
```
Should check their annotations?

jcf94 replied:
Check added, and this really helped me to find a hidden bug. :)

comaniac commented on tests/python/unittest/test_auto_scheduler_measure.py:
```diff
@@ -18,14 +18,51 @@
 
 def test_record():
-    dag, s = get_tiled_matmul()
+    A = te.placeholder((512, 512), name='A')
```
Move the checker in L67 to here. If LLVM is disabled, we don't have to even build this DAG.

comaniac commented:
> Just noticed that #6077 is also titled as "part 1", shall we make one of them part 2 instead?

Part 1 will have several PRs. "Phase 1" might be a better name to describe this proess, but anyways.

jcf94 commented:
> > Just noticed that #6077 is also titled as "part 1", shall we make one of them part 2 instead?
> 
> Part 1 will have several PRs. "Phase 1" might be a better word to describe this process, but anyways.

@junrushao1994 My bad, yes I was intended to express a meaning of stage/phase...
@comaniac Let's just use phase, I'll rename these PRs.

jcf94 commented on src/auto_scheduler/measure_record.cc:
```diff
@@ -169,6 +206,18 @@ struct Handler<::tvm::Array<::tvm::auto_scheduler::Step>> {
           fused_ids.push_back(i);
         }
         data->push_back(::tvm::auto_scheduler::FuseStep(stage_id, fused_ids));
+      } else if (name == "AN") {
```
Good point.
@merrymercy Currently to add a new transform steps, we have to modify 7 files ... Maybe it would better to move all specific implementation to transform_step.h/cc, and left the other file to just add some interface.

merrymercy replied:
I agree with your points.

jcf94 replied:
I'm working on this.

FrozenGene commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -161,16 +202,116 @@ def fuse(self, stage, iters):
+            output tensor.
+        iterator : Iterator
+            The iterator to be binded.
+        thread_name : str
+            The thread type to be binded. Currently support:
+            - vthread
+            - blockIdx.x
+            - threadIdx.x
+            - blockIdx.y
+            - threadIdx.y
```
We could add one brief comment why we don't need `blockIdx.z` / `threadIdx.z`, which is mentioned in last pr by @junrushao1994 Does it make sense to you?

junrushao1994 replied:
Sure it makes sense :-)

merrymercy replied:
We should add them. Because other policies can use them.

junrushao1994 replied:
I agree with @merrymercy that it is better to add them back

jcf94 replied:
blockIdx.z & threadIdx.z Added.

junrushao1994 commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -82,6 +82,83 @@ String ReorderStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+void ComputeAtStepNode::ApplyToSchedule(Array<te::Stage>* stages,
+                                        StageToAxesMap* stage_to_axes) const {
+  te::Stage stage = (*stages)[stage_id];
+  const IterVar& target_axis = (*stage_to_axes)[(*stages)[target_stage_id]][target_iter_id];
+  stage.compute_at((*stages)[target_stage_id], target_axis);
```
nit: make it slightly more succinct
```suggestion
  const te::Stage& target_stage = (*stages)[target_stage_id];
  const IterVar& target_axis = (*stage_to_axes)[target_stage][target_iter_id];
  stage.compute_at(target_stage, target_axis);
```

junrushao1994 commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -82,6 +82,83 @@ String ReorderStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+                                          StageToAxesMap* stage_to_axes) const {
+  auto stage = (*stages)[stage_id];
+  stage.compute_root();
+
```
nit: remove this blank line.

junrushao1994 commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -82,6 +82,83 @@ String ReorderStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+                                            StageToAxesMap* stage_to_axes) const {
+  auto stage = (*stages)[stage_id];
+  stage.compute_inline();
+
```
ditto

jcf94 commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -90,12 +90,69 @@ Stage::Stage(te::Operation op, StageKind op_type, const Array<Iterator>& iters,
+  DeleteStageEntry(pnode, stage_id);
+}
+
+void AttachMap::UpdateIters(const std::vector<IterKey>& old_iters,
+                            const std::vector<IterKey>& new_iters) {
```
We already have a vector size check in this function. :)

comaniac replied:
Ah I found it. Sorry for missing that line.

jcf94 replied:
Moved this check to front.

junrushao1994 commented on src/auto_scheduler/loop_state.h:
```diff
@@ -217,6 +196,68 @@ class Stage : public ObjectRef {
+   */
+  void DeleteStage(int stage_id);
+  /*!
+   * \brief Update the iterator relations in AttachMap.
```
I agree with Cody...Tried to understand the doc for a while but didn't succeed without looking at the source code. That will be great if we have clearer doc and method name :-)

jcf94 replied:
Doc updated.

jcf94 commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -161,16 +202,116 @@ def fuse(self, stage, iters):
+            "threadIdx.x": 6,
+            "blockIdx.y": 7,
+            "threadIdx.y": 8,
+        }
```
Moved this to class static member.

jcf94 commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -143,12 +265,67 @@ void State::DoReorderStep(const ReorderStep& step) {
+  const Stage& stage = operator->()->stages[step->stage_id];
+
+  // After compute_at, we don't know the accurate length information any more
+  // If we do want to know the accurate lengths, we can call ComputeDAG::InferBound
```
Doc updated.

jcf94 commented on tests/python/unittest/test_auto_scheduler_loop_state.py:
```diff
@@ -61,5 +61,79 @@ def test_split_fuse_reorder():
+        "                for rx (None)\n" + \
+        "                  compute = ...\n" + \
+        "    for i3 (0,112)\n" + \
+        "      compute = ...\n"
```
Updated.

jcf94 commented on src/auto_scheduler/utils.h:
```diff
@@ -89,6 +89,15 @@ inline int GetIndex(const Array<T>& array, const T& to_locate) {
 
+/*! \brief Delete the item in a std::vector. */
+template <typename T>
+inline void DeleteItem(std::vector<T>* array, const T& to_delete) {
```
Updated.

merrymercy commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -161,16 +235,116 @@ def fuse(self, stage, iters):
+        Parameters
+        ----------
+        stage : Union[int, Operation, Tensor]
+            The Stage to be vectorized, can be a Stage order index, Stage operation or stage
```
```suggestion
            The Stage to be vectorized, which can be specified by 
            the integer index, Operation, or output tensor of the stage.
```
Propagate this modification to other comments as well.

merrymercy commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -119,9 +137,61 @@ def reorder(self, stage, order):
+
+        Notes
+        -----
+        After compute_at, the extent of each iterator may not be accurate any more, so the bound
+        information will be removed from this state. Run ComputeDAG::InferBound to recover.
```
```suggestion
        After compute_at, we need careful dependency analysis to compute the accurate bound information.
        However, it is relatively expensive and complicated. So in LoopState, we just fill "None" as bound 
        for the newly created iterators. If you do need the bound, you can call ComputeDAG::InferBound on
        the returned state to get all bound information.
```

Please propagate this change.

merrymercy commented on src/auto_scheduler/compute_dag.cc:
```diff
@@ -276,10 +276,18 @@ std::pair<te::Schedule, Array<te::Tensor>> ComputeDAG::ApplySteps(
     // return value, so the ApplyToSchedule is not able to be merged to single interface
     if (auto ps = step.as<ReorderStepNode>()) {
       ps->ApplyToSchedule(stages, stage_to_axes);
+    } else if (auto ps = step.as<ComputeAtStepNode>()) {
```
Why do you use a different order from the order in our internal repo?
I propose we can use two orders
1.  Alphabetical order
2. Logical order from easy to hard.
   - primitives working on one stage: reorder, split, fuse, annotation
    - primitives working on multiple stages: compute_at, compute_root, compute_inline
   - primitive adding new stages: cache_read, cache_write, ...

And we should keep the same order in all places.

jcf94 replied:
Ok, in the .h & .py I was tring to place the functions with similar return format together.

void reorder
void compute_at
void compute_root
void compute_inline
Array<Iterator> split
Iterator fuse
Iterator vectorize
Iterator parallel
Iterator unroll
Iterator bind

It's fine to modified to the order you suggested.

merrymercy commented on tests/python/unittest/test_auto_scheduler_measure.py:
```diff
@@ -18,17 +18,55 @@
+    # Parallel
+    s.parallel(C, s[C].iters[0])
+    # Thread bind
+    s.bind(C, s[C].iters[1], "blockIdx.x")
```
Using thread binding on a llvm target is confusing to me. Although the check is correct.

jcf94 replied:
Maybe add a comment here saying this is just for testing? ...

comaniac replied:
What are we trying to test here? I didn't see a corresponding assertion.

jcf94 replied:
This test is for measure records, I tried to add different steps here to check if they can be serialized to record & reload from record successfully.

junrushao1994 commented:
A minor typo we didn't catch before (github doesn't allow me to comment on that line because it is not related to this change): in file "loop_state.h", document for class StageNode,  "Similar to te::Stage in \`include/schedule.h\`", the path is incorrect.

jcf94 commented:
Moved all the implementation to transform_step.h & transform_step.cc, and updated the order of steps.
@merrymercy @junrushao1994 @comaniac @FrozenGene

comaniac commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -51,129 +53,539 @@ const char* IteratorAnnotationString[] = {
-  node->stage_id = stage_id;
-  for (const auto& x : after_ids) {
-    CHECK(x->IsInstance<IntImmNode>());
+Step StepReadFromRecord(dmlc::JSONReader* reader) {
```
It would be better to have a step registration mechanism to establish a step lookup table from record prefix to the step node class, so that we can totally hide step node dispatching from developers. The potential issues in the current implementation is that 1) people may forget to update this function when adding a new step, and 2) there's no mechanism to check duplicated record prefix of two steps.

In addition, I feel this issue is important but I'm not sure if it's proper to address it in this PR. We may need versioning for records.

merrymercy replied:
1. Yes, we can add a registration mechanism. But I think it can be done in future PRs.
2. The current records do have a version record 
https://github.com/apache/incubator-tvm/blob/7ac78d96611ab3a069fae9eb5210e433eb240cc5/src/auto_scheduler/measure_record.cc#L314

merrymercy commented on src/auto_scheduler/loop_state.cc:
```diff
@@ -244,285 +197,73 @@ Iterator State::unroll(int stage_id, const Iterator& it, int max_unroll) {
-    } else {
-      LOG(FATAL) << "Invalid step: " << step;
-    }
+    StepApplyToState(step, this, dag);
```
1. We can remove `StepApplyToState`
2. `dag` will be used in other steps such as cache_write and rfactor.

jcf94 replied:
The `step->ApplyToState()` is actually not a virtual function(the return value of different steps may be different), so I add the `StepApplyToState` to pack all those step type check to `transform_steps.cc`.

merrymercy commented on src/auto_scheduler/measure_record.cc:
```diff
@@ -82,98 +63,23 @@ struct Handler<::tvm::Array<::tvm::auto_scheduler::Step>> {
-      s = reader->NextArrayItem();
-      CHECK(!s);
+      data->push_back(::tvm::auto_scheduler::StepReadFromRecord(reader));
+      CHECK(!reader->NextArrayItem());
```
`NextArrayItem` has side effects (i.e. moving the cursor of the reader), so I don't want to put it into the debug-oriented `CHECK`. In my opinion, the code should run correctly if we delete all `CHECK`.

jcf94 replied:
Ok, I think I'd better add the bool value s back here...

merrymercy commented on src/auto_scheduler/compute_dag.cc:
```diff
@@ -324,17 +314,9 @@ String ComputeDAG::PrintStepsAsPython(const Array<Step>& transform_steps) const
     }
   }
-  // Call each step's PrintAsPythonAPI method
+  // Call each step's ApplyToPythonAPI method
```
`PrintAsPythonAPI` is a better name in my opinion

jcf94 replied:
Recover to PrintAsPythonAPI.

jcf94 commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -51,129 +53,539 @@ const char* IteratorAnnotationString[] = {
-    if (i != after_ids.size() - 1) {
-      ss << ", ";
-    }
+void StepApplyToSchedule(const Step& step, Array<te::Stage>* stages,
```
Emm, `step->ApplyToState`, `step->ApplyToSchedule` & `step->PrintAsPythonAPI` all have problems to be written to virtual functions, we can later try to see if there's any better way to deal with this.

FrozenGene commented on src/auto_scheduler/compute_dag.cc:
```diff
@@ -326,15 +316,7 @@ String ComputeDAG::PrintStepsAsPython(const Array<Step>& transform_steps) const
-    } else {
-      LOG(FATAL) << "Invalid Step";
-    }
+    ss << StepPrintAsPythonAPI(step, &stages, &stage_to_axes);
```
ditto. PrintStepAsPythonAPI

FrozenGene commented on src/auto_scheduler/compute_dag.cc:
```diff
@@ -270,19 +270,9 @@ std::pair<te::Schedule, Array<te::Tensor>> ComputeDAG::ApplySteps(
-    } else {
-      LOG(FATAL) << "Invalid Step";
-    }
+    StepApplyToSchedule(step, stages, stage_to_axes);
```
Why we don't name it as `ApplyStepToSchedule`?

jcf94 replied:
This corresponds to each step's ApplyToSchedule, so I just name it like this ...

jcf94 commented:
> Given that the issue of 3 dispatching functions will be addressed later, this PR LGTM. Thanks for the efforts!.

Thanks, updated our merge plan to track this issue: https://github.com/merrymercy/Ansor/issues/65
